--- a/fs/fat/fs_fat32dirent.c
+++ b/fs/fat/fs_fat32dirent.c
@@ -99,11 +99,17 @@
 static inline int fat_parselfname(FAR const char **path,
                                   FAR struct fat_dirinfo_s *dirinfo,
                                   FAR char *terminator);
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_createalias(FAR struct fat_dirinfo_s *dirinfo);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_findalias(FAR struct fat_mountpt_s *fs,
                                 FAR struct fat_dirinfo_s *dirinfo);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_uniquealias(FAR struct fat_mountpt_s *fs,
                                   FAR struct fat_dirinfo_s *dirinfo);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 static int fat_path2dirname(FAR const char **path,
                             FAR struct fat_dirinfo_s *dirinfo,
@@ -119,11 +125,15 @@
                                    FAR struct fat_dirinfo_s *dirinfo);
 
 #endif
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_allocatesfnentry(FAR struct fat_mountpt_s *fs,
                                        FAR struct fat_dirinfo_s *dirinfo);
-#ifdef CONFIG_FAT_LFN
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_allocatelfnentry(FAR struct fat_mountpt_s *fs,
                                        FAR struct fat_dirinfo_s *dirinfo);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 static inline int fat_getsfname(FAR uint8_t *direntry, FAR char *buffer,
                                 unsigned int buflen);
@@ -134,18 +144,28 @@
                                 FAR struct fs_dirent_s *dir,
                                 FAR struct dirent *entry);
 #endif
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putsfname(FAR struct fat_mountpt_s *fs,
                          FAR struct fat_dirinfo_s *dirinfo);
-#ifdef CONFIG_FAT_LFN
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static void fat_initlfname(FAR uint8_t *chunk, int nchunk);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifndef CONFIG_FAT_FORCE_READONLY
 static void fat_putlfnchunk(FAR uint8_t *chunk, FAR const lfnchar *src,
                             int nchunk);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putlfname(FAR struct fat_mountpt_s *fs,
                          FAR struct fat_dirinfo_s *dirinfo);
-#endif
+#endif /* !CONFIG_FAT_FORCE_READONLY */
+#endif
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putsfdirentry(FAR struct fat_mountpt_s *fs,
                              FAR struct fat_dirinfo_s *dirinfo,
                              uint8_t attributes, uint32_t fattime);
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 
 #if defined(CONFIG_FAT_LFN) && defined(CONFIG_FAT_LFN_UTF8)
 static int fat_utf8toucs(FAR const char **str, FAR lfnchar *ucs);
@@ -669,6 +689,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_createalias(FAR struct fat_dirinfo_s *dirinfo)
 {
   uint8_t      ch;        /* Current character being processed */
@@ -893,6 +914,7 @@
 #endif
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -911,6 +933,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_findalias(FAR struct fat_mountpt_s *fs,
                                 FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -934,6 +957,7 @@
 
   return fat_findsfnentry(fs, &tmpinfo);
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -953,6 +977,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_uniquealias(FAR struct fat_mountpt_s *fs,
                                   FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -1087,6 +1112,7 @@
 
   return ret;
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -1565,6 +1591,7 @@
  *
  ****************************************************************************/
 
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_allocatesfnentry(FAR struct fat_mountpt_s *fs,
                                        FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -1639,6 +1666,7 @@
         }
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 
 /****************************************************************************
  * Name: fat_allocatelfnentry
@@ -1651,6 +1679,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static inline int fat_allocatelfnentry(FAR struct fat_mountpt_s *fs,
                                        FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -1785,6 +1814,7 @@
         }
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -2174,6 +2204,7 @@
  *
  ****************************************************************************/
 
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putsfname(FAR struct fat_mountpt_s *fs,
                          FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -2190,6 +2221,7 @@
   fs->fs_dirty = true;
   return OK;
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 
 /****************************************************************************
  * Name: fat_initlfname
@@ -2201,6 +2233,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static void fat_initlfname(FAR uint8_t *chunk, int nchunk)
 {
   int i;
@@ -2215,6 +2248,7 @@
       chunk += sizeof(uint16_t);
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -2227,6 +2261,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static void fat_putlfnchunk(FAR uint8_t *chunk, FAR const lfnchar *src,
                             int nchunk)
 {
@@ -2248,6 +2283,7 @@
       chunk += sizeof(uint16_t);
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -2261,6 +2297,7 @@
  ****************************************************************************/
 
 #ifdef CONFIG_FAT_LFN
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putlfname(FAR struct fat_mountpt_s *fs,
                          FAR struct fat_dirinfo_s *dirinfo)
 {
@@ -2463,6 +2500,7 @@
       seqno = nentries;
     }
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 #endif
 
 /****************************************************************************
@@ -2475,6 +2513,7 @@
  *
  ****************************************************************************/
 
+#ifndef CONFIG_FAT_FORCE_READONLY
 static int fat_putsfdirentry(FAR struct fat_mountpt_s *fs,
                              FAR struct fat_dirinfo_s *dirinfo,
                              uint8_t attributes, uint32_t fattime)
@@ -2504,6 +2543,7 @@
   fs->fs_dirty = true;
   return OK;
 }
+#endif /* !CONFIG_FAT_FORCE_READONLY */
 
 /****************************************************************************
  * Public Functions
