--- a/cmake/onnxruntime_common.cmake
+++ b/cmake/onnxruntime_common.cmake
@@ -86,6 +86,9 @@
 source_group(TREE ${REPO_ROOT} FILES ${onnxruntime_common_src})
 
 onnxruntime_add_static_library(onnxruntime_common ${onnxruntime_common_src})
+if(CMAKE_SYSTEM_NAME STREQUAL "NuttX")
+  target_compile_definitions(onnxruntime_common PRIVATE ORT_K7_NUTTX=1)
+endif()
 if(WIN32)
   if("cxx_std_23" IN_LIST CMAKE_CXX_COMPILE_FEATURES)
     set_property(TARGET onnxruntime_common PROPERTY CXX_STANDARD 23)
